### PR TITLE
[optim][muon] add allow_batched_matrices to Muon for [..., M, N] parameters

### PR DESCRIPTION
**Pull request by weifengpy**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #190597

* adds an opt-in `allow_batched_matrices` flag to `torch.optim.Muon` for parameters shaped `[..., M, N]`
* why allowing ndim > 2 instead of ndim == 3 exactly: a grouped MoE expert weight is 3D, [num_experts, hidden_dim, dim], and that same weight stacked across layers is 4D, [n_layers, num_experts, hidden_dim, dim]

### CODE REVIEW

File: torch/optim/_muon.py [hunk 1]
```diff
@@ -47,32 +59,49 @@ def _zeropower_via_newtonschulz(
...
-        gram_matrix = ortho_grad @ ortho_grad.T
-        gram_update = torch.addmm(
-            gram_matrix, gram_matrix, gram_matrix, beta=b, alpha=c
-        )
-        ortho_grad = torch.addmm(ortho_grad, gram_update, ortho_grad, beta=a)
+    # Fixes a latent bug: grad.bfloat16() returns grad itself when grad is
+    # already bf16, and the div_ below is in place. With nesterov=False the
+    # caller passes the momentum buffer directly, so the optimizer state was
+    # normalized on every step. copy=True never aliases, whatever the dtype.
+    ortho_grad = grad.to(dtype=torch.bfloat16, copy=True)
```
Comment (janeyx99) [inline torch/optim/_muon.py:65]: move the bug to a diff PR stacked in front of this one with a use case. no need to explain in a forever comment that a bug has been fixed

File: torch/optim/_muon.py [hunk 2]
```diff
@@ -280,6 +324,11 @@ def step(self, closure=None):
         ns_steps (int, optional): number of Newton–Schulz iteration steps. (default: {DEFAULT_NS_STEPS})
         adjust_lr_fn (str, optional): function to adjust learning rate. One of "original", "match_rms_adamw", and "spectral_unclamped".
             If not specified, we will default to use "original". (default: None)
+        allow_batched_matrices (bool, optional): opt in to parameters shaped :math:`[..., M, N]`. The last two dimensions
+            are the matrix dimensions and any leading dimensions are treated as a batch of independent matrices,
+            each orthogonalized on its own. This is useful for per-head or per-expert Muon, where a fused
+            parameter stores many logical matrices. When ``False``, only 2D parameters are accepted, so that a
+            future >2D Muon formulation is free to interpret the leading dimensions differently. (default: False)
```
Comment (janeyx99) [inline torch/optim/_muon.py]:
```suggestion
            parameter stores many logical matrices. When ``False``, only 2D parameters are accepted. (default: False)
```

File: torch/optim/_muon.py [hunk 3]
```diff
@@ -28,6 +28,18 @@
 DEFAULT_NS_STEPS = 5
 
 
+def _ndim_supported(ndim: int, allow_batched_matrices: bool) -> bool:
+    """Muon is a matrix optimizer; batches of matrices are opt-in.
+
+    Any rank above 2 is accepted rather than exactly 3, because Newton-Schulz
+    flattens every leading dimension into a single batch, so the batch rank does
+    not change the math. Transformers produce both: a grouped MoE expert weight
+    is 3D, [num_experts, hidden_dim, dim], and that same weight stacked across
```
Comment (janeyx99) [inline torch/optim/_muon.py]: how would a 3D grouped MoE expert weight be handled for muon without batch? is the first num_experts dim effectively a batch_size?

Comment (weifengpy) [reply torch/optim/_muon.py]: first num_experts is indeed a batch_size

File: torch/optim/_muon.py (hunk 1 above)
Comment (weifengpy) [reply torch/optim/_muon.py:65]: will do

Comment (weifengpy) [reply torch/optim/_muon.py:65]: the bug fix is moved into a separate PR now: https://github.com/pytorch/pytorch/pull/194664

File: torch/optim/_muon.py (hunk 2 above)
Comment (weifengpy) [reply torch/optim/_muon.py]: done

**Review by janeyx99 — COMMENTED** [review]
Overall approach looks correct. I'd like to see some perf benchmarks showcasing the batching being faster when compared to not batching (having each param be separate). I'd also like to see documentation for how we expect someone will use allow_batched_matrices.

One question I have is why we don't collapse the addmm branch with the baddmm branch where b=1. 

Additionally, Claude brought up some good points:

**1. Loading a pre-PR Muon checkpoint will now raise `KeyError`.**
  Adding `allow_batched_matrices` to `defaults` means it's read from the param
  group at `_muon.py:236`, but `Optimizer.load_state_dict`'s `update_group`
  (`optimizer.py:1031-1041`) returns the *saved* group dict verbatim apart from
  `params`/`param_names`. A checkpoint written before this PR therefore produces
  a group with no `allow_batched_matrices` key. The usual fix is a `__setstate__`
  backfill (`sgd.py:75-82` is the canonical shape)

**2. The new `optim_inputs_func_muon` entry doesn't reach the batched path.**
  `OptimizerInput(params=None, kwargs={"allow_batched_matrices": True})` lets the
  generic suite supply its own params, and `test_optim.py` hardcodes 2D params
  for Muon in ~9 places. So the `ndim > 2` branch in `_zeropower_via_newtonschulz` is
  never taken by this input. That also means the batched path gets nothing from the generic machinery. Ideally we would expand the test coverage so that the batched matrices are tested as well.
